### ISSUE TITLE
Client connection abstraction

### ISSUE THREAD

**Issue by aelsi2** [Scrap-Mechanic-Modding/SM_Server#1]
I suggest adding an interface for exchanging packets with clients:
```cs
public interface IClientConnection : IDisposable
{
    public void ReceivePacket<T>(byte packetId, Action<IClient, T> handler) where T : new, IPacket;
    public void SendPacket<T>(T packet) where T : new, IPacket;
}
```
An implementation of this interface that uses `Steamworks.SocketManager` could be called `SteamworksConnection`.

The `new` constraint and the `packetId` parameter in `IClientConnection.ReceivePacket<T>` would make it possible to match, create and deserialize packets without using reflection. An extension method could also be added to allow you to subscribe to incoming packets without specifying the `packetId` that would get it through reflection.

**Comment by arcturus-prime** [comment]
Removed the reflection for packet IDs entirely and added the `new` constraint with commit 9c8fd5a and 412dc44. We don't really need the extension method, and we'd like to use the type, so we neglected to add that.

**Comment by aelsi2** [comment]
412dc44:
> No more reflection!

There's still reflection though (`SmartSocket.OnMessage`):
```cs
dynamic packet = Activator.CreateInstance(packetType);
//...
Type[] typeArgs = { typeof(Connection), packetType };
var callbackType = typeof(Action<,>).MakeGenericType(typeArgs);
//etc.
```

I was going to open a PR and close the issue myself.

**Comment by aelsi2** [comment]
My thoughts were:
Add a field with the list of handlers by ids like this:
```cs
private readonly List<(byte packetId, Action<Connection, BinaryReader> handler)> handlers;
```
Make `ReceivePacket` add a delegate to the list:
```cs
public void ReceivePacket<T>(byte packetId, Action<Connection, T> handler) where T : IPacket, new()
{
    Action<Connection, BinaryReader> outerHandler = (connection, reader) =>
    {
        var packet = new T();
        packet.Deserialize(reader);
        handler(connection, packet);
    };
    handlers.Add((packetId, outerHandler));
}
```
Then in `OnMessage` iterate throught the list and call the handlers with the matching `packetId`:
```cs
// ...
foreach (var (curPid, callback) in callbacks)
{
    if (curPid == 0 || curPid == id)
    {
        callback(connection, dReader);
        dStream.Position = 0;
    }
}
```
An extension method would simplify the subscription:
```cs
public static void ReceivePacket<T>(this SmartSocket socket, Action<Connection, T> handler) where T : IPacket, new()
{
    byte packetId =  typeof(T).GetProperty(nameof(IPacket.PacketId), BindingFlags.Static | BindingFlags.Public)
        ?.GetValue(null) as byte? ?? throw new NotSupportedException();
    socket.ReceivePacket(packetId, handler);
}
```

**Comment by aelsi2** [comment]
I've just realized [static virtual members](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/static-virtual-interface-members) were a thing so I guess the extension is really not needed.

**Comment by aelsi2** [comment]
Closed by 69df77f617d6449214e6e7d704e6c38c76d40d46
